ETL-2219.  Changes to common_source.sh and ib_dataload.sh for the new ETL repo

hema1024 opened the pull request:
ETL-2219: Changes to common_source.sh / ib_dataload.sh for new repo

* removing the need for username.ini in the $HOME dir
* processes can run from any dir it is released to, /home/${username}/ is not mandatory
* split the methods in common_source.sh into multiple modules under common/bin/
* moved mart_helper_methods.sh, impala_helper_methods.sh, mart.conf into its own library called 'aggregation_helper'
* global conf files: prod.inf, staging.inf, dev.inf will reside in the $HOME dir on production and etlscripts/ dir in the git repo.  thresher->prod.inf, etlstage->staging.inf, all others->dev.inf will be loaded by common_source.sh when sourced in.  Also added support for loading thresher.inf if prod.inf is not present (this is to support processes being released to existing etl servers where global config file is called thresher.inf)
* The following global variables will be set by common_source.sh depending on runtime profile and etl environment
ETL_ENVIRONMENT (value will be one of prod/dev/staging), PROCESS_BINDIR
PROCESS_CONFDIR (will be set depending on runtime profile and etl environment), PROCESS_PIGDIR, PROCESS_SQLDIR, PROCESS_LIBDIR, APPNAME (this is for backward compatibility, will contain EXECUSER value) 
LOGDIR, TMPDIR, PIDDIR, 
COMMON_LIB_BASEDIR, COMMON_LIB_BINDIR 
* ib_dataload.sh has been changed to work with the new changes

Test runs for dev and prod settings: 
https://util01.303net.net/jira/secure/attachment/42424/ib_dataload_201509.log

ETL-2240: Added a pom.xml for ib dataload builds
* Will take runtime_profile command line parameter and bundle the appropriate runtime config files into a zip file.
* Added validation for the above.  'runtime_profile' has to be supplied
* The final zip file will contain artifact version and runtime profile name
* It will include a version.txt with project name, version, runtime profile, build date time

Test runs:
https://util01.303net.net/jira/secure/attachment/42425/build_test.log

hema1024 commented on aggregation_helper/bin/impala_helper_methods.sh:
```diff
@@ -0,0 +1,366 @@
+#!/bin/bash
```
no change in this file, just put a copy from p-etl03 repo

hema1024 commented on aggregation_helper/bin/mart_helper_methods.sh:
```diff
@@ -0,0 +1,872 @@
+PIGLOGDIR="${LOGDIR}/log/sty"
+PIGPARAMS="-Dmapred.reduce.slowstart.completed.maps=0.5"
+
+source $AGG_HELPER_LIB_CONFDIR/mart.conf
```
no change to this file, except this line where the path of the conf file is now base on the new logic in common_source.sh

hema1024 commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,785 @@
+#!/bin/bash
+
+##############################################
```
my changes to common_source.sh starts here ...

hema1024 commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,785 @@
+        echo "${PROCESS_BASEDIR}/${LIB_NAME}/conf/${ETL_ENVIRONMENT}"
+    fi
+}
+
```
first part of the changes ends here.  After this is copy pasted code for PID related stuff

hema1024 commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,785 @@
+   log "Start    :: $SCRIPT"
+
+   # the inf / configuration files get sourced below!
+
```
from this line forward up to the end of this method are changes for sourcing in global config and process config depending on ETL_ENVIRONMENT and RUNTIME_PROFILE

hema1024 commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,785 @@
+    fi
+}
+
+
```
new changes for setting locale, username variables, etl env, global & process paths & library paths

hema1024 commented on common/bin/date_util.sh:
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
methods in common_source.sh organized into multiple files

hema1024 commented on ib_dataload/bin/ib_dataload.sh:
```diff
@@ -0,0 +1,177 @@
+#!/bin/bash
```
lines 3, 112-124 are new changes, rest is same as old code

ibogomazov commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,791 @@
+function set_locale() {
+    export LANG=en_US.UTF-8
+    export LANGUAGE=en
+    export LC_CTYPE=en_US.UTF-8
```
don't you want it to be LC_ALL?

ibogomazov commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,791 @@
+# the global conf dir is the location of prod.inf / staging.inf / dev.inf
+# and it is expected to be in one dir above the process basedir 
+function set_global_conf_path() {
+    GLOBAL_CONFDIR=$(echo ${PROCESS_BASEDIR} | rev | cut -d'/' -f2- | rev)
```
the result will be surprisingly different for /home/thresher/bin/ and /home/thresher/bin, maybe just GLOBAL_CONFDIR=${PROCESS_BASEDIR}/../ ?

hema1024 replied:
ok, fixed using 
echo ${PROCESS_BASEDIR%/} | rev | cut -d'/' -f2- | rev

ibogomazov commented on ib_dataload/bin/ib_dataload.sh:
```diff
@@ -0,0 +1,177 @@
+fi
+
+start_script
+create_pidfile
```
I believe command line parameters should go after create_pidfile invocation

hema1024 replied:
This script needs the RUNTIME_PROFILE value to be set before sourcing in common_source...so I moved it here.

hema1024 commented on common/bin/log_collection_util.sh:
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
```
moved the log collection methods in common_source.sh into this file, no code changes

hema1024 commented on common/bin/sql_util.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
moved the sql related methods in common_source.sh into this file, no code changes

hema1024 commented on common/bin/trafficviewer_util.sh:
```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
```
moved the traffic viewer methods in common_source.sh into this file, no code changes

hema1024 commented on common/bin/verify_agg_util.sh:
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
```
moved the is_agg_* check methods in common_source.sh into this file, no code changes

dejiang-jin replied:
should not it belong to aggregation_helper?

hema1024 replied:
good point, moved

hema1024 commented on dev.inf:
```diff
@@ -0,0 +1,267 @@
+################################################
```
no changes, put a copy from p-etl03

hema1024 commented on ib_dataload/assembly/ib_dataload.xml:
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
this is new code.  an assembly file for packaging all files needed for ib_dataload release

hema1024 commented on ib_dataload/pom.xml:
```diff
@@ -0,0 +1,82 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
new pom.xml for building ib_dataload artifacts for staging and production

hema1024 commented on prod.inf:
```diff
@@ -0,0 +1,283 @@
+################################################
```
copied thresher.inf from p-etl03 and renamed as prod.inf.
Before the actual release, this file will be updated again.
The idea is to remove thresher.inf from p-etl03 repo, and release it from the new repo to all p-etl servers using a jenkins job.

dejiang-jin commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,791 @@
+    COMMON_LIB_BINDIR=$COMMON_LIB_BASEDIR/bin
+
+    # set paths for 'aggregation_helper' library
+    AGG_HELPER_LIB_BASEDIR=$PROCESS_BASEDIR/${AGG_HELPER_LIB_NAME}
```
does the aggregation_helper module as part of common, is used by all other modules?
if not, we may not need to define it here

hema1024 replied:
removed agg_helper defs from here

dejiang-jin commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,791 @@
+    if ! is_dev_user || [[ -z "${RUNTIME_PROFILE}" ]]; then
+        echo "${PROCESS_BASEDIR}/conf"
+    else
+        echo "${PROCESS_BASEDIR}/runtime/${RUNTIME_PROFILE}/${ETL_ENVIRONMENT}/conf"
```
it may not be relevant. in ib_dataload/pom.xml 
<message>etl_environment property must be set to one of 'prod'	or 'staging'}</message>
I wonder what's the value for ETL_ENVIRONMENT and RUNTIME_PROFILE here (if is_dev_user == true )
Should we "${PROCESS_BASEDIR}/conf" simply? ( let install program pull corresponding configuration to here base on info in pom)

hema1024 replied:
Per discussion in the last meeting, I am leaving config files for dev/staging/prod under conf dir, and let common_source.sh source in appropriate file depending on the executing user.

ffan07039 commented:
Hema,
I only see one non-blocking issue:
At release stage, the artifact is designated by '*etl_environment*' and it's stored in **version.txt**
At run time, '*etl_environment*' is determined by current user name. Say in function **get_etl_enfironment**, can we double check the **version.txt** so that the script will immediately fail it doesn't match?
Feng

ffan07039 commented:
:+1:

ibogomazov commented:
@hema1024 which commit did you take as a base for this branch? I hope it will be much clearer to look at the diff with that one

hema1024 commented:
@ibogomazov @ffan07039 I'm making some more changes after the meeting couple of days ago.  It was decided that
* conf dirs for dev/prod/staging will remain in the final build for all environments, instead bundling these separately during release process.  This is so we dont have to make separate builds for staging and production.  The same artifact that was tested on staging can be released to production.
* some kind of versioning for common lib
I'll update the PR.

hema1024 commented:
@ibogomazov the base for this PR is an empty master branch.  Mine is the first commit in this repo.

hema1024 commented:
@ffan07039 @ibogomazov I've checked in my changes, can you please review.
Changes from your last review
* Removed etl_environment input to pom.xml, so all config files (dev/staging/prod) will be bundled in the final zip, and the appropriate config file will be loaded depending on the executing user.  Updated common_source.sh and ib_dataload.sh for this logic.
* Removed AGG_HELPER_LIB variables creation inside common_source.sh

ffan07039 commented:
:+1:  again!

ibogomazov commented on common/bin/common_source.sh:
```diff
@@ -0,0 +1,779 @@
+    fi
+}
+
+function source_file_if_exists() {
```
what is the purpose of this function?
source won't interrupt script execution unless errexit enabled, as well as a diagnostic message is sent to stderr if file does not exist

ibogomazov replied:
maybe another one would be helpful source_file_or_halt()?

hema1024 replied:
Igor, the reason I created this method is: common_source.sh is now sourcing global and process related config files depending on executing user.  I wanted a method that can log the path details of sourced file, if it was successfully sourced, or the file was not found.  I created a method to avoid code repetition in multiple places.

hema1024 replied:
yes source_file_or_halt will be helpful, I'll create one when we need it.  Right now we dont want to stop execution of the script if the config files are not found because not all processes have config files.

ibogomazov commented:
:+1:

hema1024 commented:
@ffan07039 @ibogomazov thanks for the review.
